src/Diagnostics.DataProviders/KustoHeartBeatService/KustoHeartBeatService.cs

```diff
@@ -0,0 +1,183 @@
+
+        public KustoHeartBeat(string primaryCluster, string failoverCluster, KustoDataProvider kustoDataProvider, KustoDataProviderConfiguration configuration)
+        {
+            PrimaryCluster = primaryCluster;
+            FailoverCluster = failoverCluster;
+            UsePrimaryCluster = true;
+            _kustoDataProvider = kustoDataProvider;
+            _configuration = configuration;
+        }
+
+        public async Task RunHeartBeatTask(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested) {
+                bool primaryHeartBeatSuccess = false;
+                bool failoverHeartBeatSuccess = false;
+                Exception exceptionForLog = null;
+                string activityId = Guid.NewGuid().ToString();
+                try
+                {
+                    // Run kusto query with time out
+                    var primaryTask = _kustoDataProvider.ExecuteQueryForHeartbeat(_configuration.HeartBeatQuery, PrimaryCluster, _configuration.HeartBeatTimeOut, activityId);
+                    var failoverTask = _kustoDataProvider.ExecuteQueryForHeartbeat(_configuration.HeartBeatQuery, FailoverCluster, _configuration.HeartBeatTimeOut, activityId);
+                    await Task.WhenAll(new Task[] { primaryTask, failoverTask });
+
+                    var result = await primaryTask;
+                    if (result.Rows.Count >= 1)
+                    {
+                        primaryHeartBeatSuccess = true;
+                    }
+
+                    var failoverResult = await failoverTask;
+                    if (failoverResult.Rows.Count >= 1)
+                    {
+                        failoverHeartBeatSuccess = true;
+                    }
+
+                }
+                catch (Exception ex)
+                {
+                    exceptionForLog = ex;
+                }
+
+                if (primaryHeartBeatSuccess)
+                {
+                    _ConsecutiveFailureCount = 0;
+                    _ConsecutiveSuccessCount++;
+                }
+                else
+                {
+                    _ConsecutiveFailureCount++;
+                    _ConsecutiveSuccessCount = 0;
+                }
+
+                // if not in failover state
+                //  should failover?
+                if (UsePrimaryCluster && _ConsecutiveFailureCount >= _configuration.HeartBeatConsecutiveFailureLimit)
+                {
+                    UsePrimaryCluster = false;
+                } // else should stop failover
+                else if (!UsePrimaryCluster && _ConsecutiveSuccessCount >= _configuration.HeartBeatConsecutiveFailureLimit)
+                {
+                    UsePrimaryCluster = true;
+                }
+
+                LogHeartBeatInformation(primaryHeartBeatSuccess, failoverHeartBeatSuccess, PrimaryCluster, FailoverCluster, activityId, _ConsecutiveSuccessCount, _ConsecutiveFailureCount, exceptionForLog);
```

Comment:
How about also logging the startTime and endTime for the activity? 
It will give you insights on the performance of this task in production.

Reply:
Since KustoClient.cs already logs time taken we could just cross reference the Activation Id with the Request Id. I can add redundant logging if cross referencing the log information is too difficult?

Reply:
I see.. good point.

Reply:
its just a nice to have logging. But you are right. we can figure that out using Kusto logging also.